homeassistant/components/mqtt/valve.py

```diff
@@ -270,6 +270,11 @@ def _process_position_valve_update(
                     msg.topic,
                 )
             else:
+                if state is None:
+                    if percentage_payload == 0:
+                        state = STATE_CLOSED
+                    elif percentage_payload == 100:
+                        state = STATE_OPEN
```

Comment:
This is already handled by `ValveEntity.state` why does MQTT need to worry about it?

Reply:
It seems it is not.

Reply:
The `ValveEntity.state` leans on the `entity._attr_is_opening` and `entity._attr_is_closing`. We need to reset those if the valve is `closed` or fully `open`. As we receive the `position` we should reset these attributes when we know the valve is fully open or closed, based on the position.
Alternatively we could also allow `open` and `closed` states here.

Reply:
This explanation doesn't make sense to me.

A valve which reports its position has the `open` and `closed` states calculated by `ValveEntity.state`, the valve's `is_closed` property is ignored.

Similarly, in `async_set_valve_position`, there's this calculation which is both wrong and uncessary:
```py
            self._update_state(
                STATE_CLOSED
                if percentage_position == self._config[CONF_POSITION_CLOSED]
                else STATE_OPEN
            )
```

IMO, the problem is the method `_update_state` which is called for both binary and position aware valves.
Split it in two instead:
```
def _update_binary_state(self, state: str) -> None:
    # state can be open, closed, opening, closing
```

```
def _update_position_state(self, state: str | None) -> None:
    # state can be opening, closing or None
```

Reply:
`update_state` is only called when an `opening` or `closing` state is received. Further the entity state is always written if an attribute was changed.

Reply:
Right, we do not have to update `self._attr_is_closed`. Fixed that